Optimize nullable value type constant comparison codegen

jjonescz opened the pull request:
Fixes part of https://github.com/dotnet/roslyn/issues/66431.
Fixes C# side of https://github.com/dotnet/roslyn/issues/52629.

Note that `is` patterns are not addressed in this PR to keep it concise.

<details>
<summary>

### Simple microbenchmark ([source](https://github.com/jjonescz/RoslynIssue66431))

</summary>

```log
BenchmarkDotNet=v0.13.5, OS=Windows 11 (10.0.22621.1555/22H2/2022Update/SunValley2), VM=Hyper-V
Intel Xeon Platinum 8370C CPU 2.80GHz, 1 CPU, 16 logical and 8 physical cores
.NET SDK=8.0.100-preview.4.23218.17
  [Host]     : .NET 8.0.0 (8.0.23.21704), X64 RyuJIT AVX2
  Job-POUQAN : .NET 8.0.0 (8.0.23.21704), X64 RyuJIT AVX2
```

| Method |    b |      Mean | Ratio | Code Size |
|------- |----- |----------:|------:|----------:|
| Before | True | 0.0270 ns |  1.00 |      30 B |
|  After | True | 0.0043 ns |  0.15 |      11 B |

```assembly
; Program.Before(System.Nullable`1<Boolean>)
; 	public bool Before(bool? b) => b == true;
; 	                               ^^^^^^^^^
       mov       [rsp+10],rdx
       movzx     eax,byte ptr [rsp+10]
       movzx     edx,byte ptr [rsp+11]
       cmp       dl,1
       sete      dl
       movzx     edx,dl
       movzx     eax,al
       and       eax,edx
       ret
; Total bytes of code 30
```

```assembly
; Program.After(System.Nullable`1<Boolean>)
; 	public bool After(bool? b) => b.GetValueOrDefault();
; 	                              ^^^^^^^^^^^^^^^^^^^^^
       mov       [rsp+10],rdx
       movzx     eax,byte ptr [rsp+11]
       ret
; Total bytes of code 11
```

</details>

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_BinaryOperator.cs:
```diff
@@ -973,6 +973,36 @@ private BoundExpression MakeNullableHasValue(SyntaxNode syntax, BoundExpression
+                        syntax: syntax,
+                        operatorKind: kind.Unlifted(),
+                        loweredLeft: xNonNull,
+                        loweredRight: MakeOptimizedGetValueOrDefault(syntax, loweredRight),
```
>MakeOptimizedGetValueOrDefault(syntax, loweredRight) [](http://example.com/codeflow?start=38&length=52)

```yNonNull ?? MakeOptimizedGetValueOrDefault(syntax, loweredRight)```?

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullCoalescingOperator.cs:
```diff
@@ -141,6 +141,18 @@ public override BoundNode VisitNullCoalescingOperator(BoundNullCoalescingOperato
+                && TryGetNullableMethod(rewrittenLeft.Syntax, rewrittenLeft.Type, SpecialMember.System_Nullable_T_get_HasValue, out MethodSymbol hasValue))
+            {
+                return _factory.Binary(BinaryOperatorKind.Or, rewrittenRight.Type,
+                    BoundCall.Synthesized(rewrittenLeft.Syntax, rewrittenLeft, getValueOrDefault2),
```
>BoundCall.Synthesized(rewrittenLeft.Syntax, rewrittenLeft, getValueOrDefault2) [](http://example.com/codeflow?start=20&length=78)

Don't we have helpers like ```MakeOptimizedGetValueOrDefault``` and for HasValue as well?

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenOptimizedNullableOperators.cs:
```diff
@@ -2414,5 +2414,562 @@ .maxstack  2
+            var expectedOutput = "1100-0011-011 0011-1100-100 0000-1111-100";
+            var verifier = CompileAndVerify(source, options: TestOptions.ReleaseExe, expectedOutput: expectedOutput);
+            verifier.VerifyDiagnostics();
+            verifier.VerifyMethodBody("C.M", """
```
>verifier.VerifyMethodBody("C.M", """ [](http://example.com/codeflow?start=12&length=36)

I understand that you are trying to rely on sequence points to match portions of IL to source code. However, this approach is fragile because any change at the beginning of the method will cause changes all the way through the rest of the IL and it will be very hard to see what changes are "interesting" and which are not. Reviewing IL for long methods (even with sequence points) is a major pain. So, please split the big method into several focused methods and assert IL for each method separately (dropping the sequence points, if it is not your goal to test them). Separating scenarios into individual unit-tests will be even better, this way a failure will not affect tests for scenarios that aren't broken.
This comment applies to other added tests too.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullCoalescingOperator.cs:
```diff
@@ -141,6 +141,18 @@ public override BoundNode VisitNullCoalescingOperator(BoundNullCoalescingOperato
+            {
+                return _factory.Binary(BinaryOperatorKind.Or, rewrittenRight.Type,
+                    BoundCall.Synthesized(rewrittenLeft.Syntax, rewrittenLeft, getValueOrDefault2),
+                    _factory.Not(BoundCall.Synthesized(rewrittenLeft.Syntax, rewrittenLeft, hasValue)));
```
>rewrittenLeft [](http://example.com/codeflow?start=77&length=13)

Are we going to evaluate ```rewrittenLeft``` twice?

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleEqualityTests.cs:
```diff
@@ -534,43 +534,22 @@ static bool M((int?, bool?) t1)
+  IL_000f:  bne.un.s   IL_001e
+  IL_0011:  ldloca.s   V_0
+  IL_0013:  ldflda     ""bool? System.ValueTuple<int?, bool?>.Item2""
+  IL_0018:  call       ""bool bool?.GetValueOrDefault()""
```
Is this a breaking change for a scenario when ```bool?``` value is 2, for example?

jjonescz replied:
I believe not. See https://github.com/dotnet/roslyn/issues/66431#issuecomment-1822751507.

AlekseyTs replied:
> I believe not.

I am not so sure about that.

jjonescz replied:
Edit: yes, you're right, it's a breaking change because the result will be different, but only for `bool`s represented differently than 0 or 1 (as discussed in the issue).

CyrusNajmabadi replied:
Note: I would not describe this as a breaking change. It is simply depending on undefined behavior.

AlekseyTs commented:
Done with review pass (commit 4), new tests in CodeGenOptimizedNullableOperatorTests were not reviewed.

DoctorKrolic commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullCoalescingOperator.cs:
```diff
@@ -141,6 +141,18 @@ public override BoundNode VisitNullCoalescingOperator(BoundNullCoalescingOperato
             }
 
+            // Optimize `left ?? true` to `left.GetValueOrDefault() | !left.HasValue`.
+            if (rewrittenRight.ConstantValueOpt == ConstantValue.True
```
I've implemented more optimal `left.GetValueOrDefault(true)` lowering for this case in https://github.com/dotnet/roslyn/pull/70199. It is more generic (applies not only to booleans), results in smaller IL thus is very good for JIT inlining. I think, we want to get #70199 first and then adjust this PR based on that

jjonescz commented:
@AlekseyTs @dotnet/roslyn-compiler this is ready for reviews, thanks

AlekseyTs commented:
Do we want to make similar changes for VB?

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenOptimizedNullableOperators.cs:
```diff
@@ -2414,5 +2414,371 @@ .maxstack  2
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/66431")]
+        public void NullableConstant_Int_GE1()
+        {
+            NullableConstant_Int("x >= -1", "1110", """
```
>NullableConstant_Int [](http://example.com/codeflow?start=12&length=20)

I am not going to block on this, but, I think the tests  would be easier to read and maintain if we were not using helpers like that, but instead had each unit-test explicitly test its own scenario as other tests in this file do. I asked to split up a single monolithic unit test into a series of unit-tests, each covering one specific scenario. The motivation for suggestion was better readability and ease of maintenance. I don't feel like I got what I was asking for. I feel like we went from one obscure way of testing things to another, perhaps, even more obscure way of testing things.

AlekseyTs commented:
It looks like at commit 6 this PR is fixing only #52629. Is this correct? Consider adjusting the description accordingly. Also, it looks like all added tests are referencing a different issue.

AlekseyTs commented:
Done with review pass (commit 6)

jjonescz commented:
> Do we want to make similar changes for VB?

Perhaps that can be done in a separate PR?

AlekseyTs commented:
I have no objection to a separate PR. Are we going to keep the issue active until then, or open a separate issue?

jjonescz commented:
> I have no objection to a separate PR. Are we going to keep the issue active until then, or open a separate issue?

I will keep the issue open. Added comment to it and changed description of this PR accordingly.

AlekseyTs commented:
@jjonescz It looks like there are legitimate test failures.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_BinaryOperator.cs:
```diff
@@ -993,6 +993,38 @@ private BoundExpression MakeNullableHasValue(SyntaxNode syntax, BoundExpression
+            TypeSymbol boolType = _compilation.GetSpecialType(SpecialType.System_Boolean);
+
+            // Optimization: If one side is non-default constant, checking HasValue is not needed.
+            if (kind.Operator() is BinaryOperatorKind.Equal or BinaryOperatorKind.NotEqual)
```
>kind.Operator() is BinaryOperatorKind.Equal or BinaryOperatorKind.NotEqual [](http://example.com/codeflow?start=16&length=74)

We need to make sure we cover all other binary operators. I do not think we need to cover intersection with all possible argument types, but, perhaps, use distinct type for each operator, if we have to add tests,

AlekseyTs replied:
In the new tests I see tests for `<` and `>=`. I think we should add tests for `<=` and `>` as well.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_BinaryOperator.cs:
```diff
@@ -993,6 +993,38 @@ private BoundExpression MakeNullableHasValue(SyntaxNode syntax, BoundExpression
+            // Optimization: If one side is non-default constant, checking HasValue is not needed.
+            if (kind.Operator() is BinaryOperatorKind.Equal or BinaryOperatorKind.NotEqual)
+            {
+                if (xNonNull?.ConstantValueOpt is { IsDefaultValue: false })
```
>xNonNull?.ConstantValueOpt is { IsDefaultValue: false } [](http://example.com/codeflow?start=20&length=55)

We need to make sure we have tests ```xNonNull``` as "default" constant value.

AlekseyTs replied:
It looks like we have some tests for this scenario

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_BinaryOperator.cs:
```diff
@@ -993,6 +993,38 @@ private BoundExpression MakeNullableHasValue(SyntaxNode syntax, BoundExpression
+                        constrainedToTypeOpt: null);
+                }
+
+                if (yNonNull?.ConstantValueOpt is { IsDefaultValue: false })
```
>yNonNull?.ConstantValueOpt is { IsDefaultValue: false } [](http://example.com/codeflow?start=20&length=55)

We need to make sure we have tests ```yNonNull``` as "default" constant value.

AlekseyTs replied:
It looks like we have some tests for this scenario

AlekseyTs reviewed (approved):
LGTM (commit 10)

jjonescz commented:
@dotnet/roslyn-compiler for a second review, thanks

jjonescz commented:
@dotnet/roslyn-compiler for a second review, thanks
